File: ios/Sources/App/Services/PortForwardingService.swift
import Foundation
import Citadel
import NIOCore

/// Port forwarding stub - Citadel doesn't currently expose port forwarding APIs.
/// This service compiles but doesn't actually create tunnels.
/// Future enhancement: Implement via raw SSH channel operations if Citadel adds support.
actor PortForwardingService {
    static let shared = PortForwardingService()

    private var activeForwards: [UUID: PortForward] = [:]

    struct PortForward: Identifiable {
            let id = UUID()
            let type: ForwardType
            let localPort: Int
            let remoteHost: String?
            let remotePort: Int?
            let sshClient: SSHClient
            var task: Task<Void, Never>?
            var isRunning: Bool = true
        }

    enum ForwardType: String, Codable {
        case local = "Local"
        case dynamic = "Dynamic (SOCKS)"
        case remote = "Remote"
    }

    /// Start a local port forward: localhost:localPort -> remoteHost:remotePort
    /// Currently a stub - Citadel doesn't expose this API
    func startLocalForward(
        connectionID: UUID,
        localPort: Int,
        remoteHost: String,
        remotePort: Int,
        sshClient: SSHClient
    ) async throws {
        throw PortForwardingError.notSupported
    }

    /// Start a dynamic (SOCKS) port forward: localhost:localPort acts as SOCKS proxy
    /// Currently a stub - Citadel doesn't expose this API
    func startDynamicForward(
        connectionID: UUID,
        localPort: Int,
        sshClient: SSHClient
    ) async throws {
        throw PortForwardingError.notSupported
    }

    /// Start a remote port forward: remoteHost:remotePort -> localhost:localPort
    /// Currently a stub - Citadel doesn't expose this API
    func startRemoteForward(
        connectionID: UUID,
        remoteHost: String,
        remotePort: Int,
        localHost: String,
        localPort: Int,
        sshClient: SSHClient
    ) async throws {
        throw PortForwardingError.notSupported
    }

    /// Stop a specific port forward
    func stopForward(forwardID: UUID) async {
        if let forward = activeForwards.removeValue(forKey: forwardID) {
            forward.task?.cancel()
        }
    }

    /// Stop all port forwards for a connection
    func stopAllForConnection(connectionID: UUID) async {
        // Stub - no actual forwards to stop
    }

    /// Get all active port forwards
    func getActiveForwards() -> [PortForward] {
        Array(activeForwards.values)
    }
}

enum PortForwardingError: Error, LocalizedError {
    case notSupported
    case connectionFailed

    var errorDescription: String? {
        switch self {
        case .notSupported:
            return "Port forwarding not yet supported by Citadel SSH library"
        case .connectionFailed:
            return "Failed to establish port forward"
        }
    }
}

File: android/app/src/main/kotlin/dev/termvault/app/ssh/PortForwardingService.kt
package dev.termvault.app.ssh

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import net.schmizz.sshj.connection.channel.direct.LocalPortForwarder
import net.schmizz.sshj.connection.channel.direct.Parameters
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

enum class ForwardType { LOCAL, DYNAMIC, REMOTE }

data class PortForward(
    val id: UUID = UUID.randomUUID(),
    val connectionId: UUID,
    val type: ForwardType,
    val localPort: Int,
    val remoteHost: String?,
    val remotePort: Int?,
)

class PortForwardingNotSupportedException(type: ForwardType) :
    Exception(
        when (type) {
            ForwardType.DYNAMIC -> "Dynamic (SOCKS) forwarding isn't implemented yet."
            ForwardType.REMOTE -> "Remote forwarding isn't implemented yet."
            ForwardType.LOCAL -> "Local forwarding failed to start."
        }
    )

/**
 * Local (`-L`) port forwarding, genuinely implemented via sshj's
 * `newLocalPortForwarder` — unlike iOS, where Citadel doesn't expose a
 * forwarding API at all and the feature is a UI-visible "not supported"
 * stub. Dynamic (SOCKS) and remote (`-R`) forwarding are left unimplemented
 * here too for now — real, scoped follow-up work, not faked.
 */
class PortForwardingService(private val sessionManager: SshSessionManager) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val forwards = ConcurrentHashMap<UUID, PortForward>()
    private val sockets = ConcurrentHashMap<UUID, ServerSocket>()
    private val jobs = ConcurrentHashMap<UUID, Job>()

    fun activeForwards(): List<PortForward> = forwards.values.toList()

    suspend fun startLocalForward(connectionId: UUID, localPort: Int, remoteHost: String, remotePort: Int): PortForward {
        val client = sessionManager.session(connectionId) ?: throw SshNotConnectedException()
        val serverSocket = ServerSocket()
        serverSocket.bind(InetSocketAddress("127.0.0.1", localPort))

        val parameters = Parameters("127.0.0.1", localPort, remoteHost, remotePort)
        val forwarder = client.newLocalPortForwarder(parameters, serverSocket)

        val forward = PortForward(
            connectionId = connectionId,
            type = ForwardType.LOCAL,
            localPort = localPort,
            remoteHost = remoteHost,
            remotePort = remotePort,
        )
        forwards[forward.id] = forward
        sockets[forward.id] = serverSocket
        jobs[forward.id] = scope.launch {
            runCatching { forwarder.listen() }
            stopForward(forward.id)
        }
        return forward
    }

    suspend fun startDynamicForward(connectionId: UUID, localPort: Int): PortForward {
        throw PortForwardingNotSupportedException(ForwardType.DYNAMIC)
    }

    suspend fun startRemoteForward(connectionId: UUID, remoteHost: String, remotePort: Int, localPort: Int): PortForward {
        throw PortForwardingNotSupportedException(ForwardType.REMOTE)
    }

    fun stopForward(forwardId: UUID) {
        jobs.remove(forwardId)?.cancel()
        runCatching { sockets.remove(forwardId)?.close() }
        forwards.remove(forwardId)
    }

    fun stopAllForConnection(connectionId: UUID) {
        forwards.values.filter { it.connectionId == connectionId }.forEach { stopForward(it.id) }
    }
}
